mlflow/tensorflow.py

```diff
@@ -13,76 +13,213 @@
-
-            # Determining output tensors.
-            fetch_mapping = {sigdef_output: graph.get_tensor_by_name(tnsr_info.name)
-                             for sigdef_output, tnsr_info in sig_def.outputs.items()}
 
-            # Build the feed dict, mapping input tensors to DataFrame column values.
-            # We assume that input arguments to the signature def correspond to DataFrame column
-            # names
-            feed_dict = {graph.get_tensor_by_name(tnsr_info.name): df[sigdef_input].values
-                         for sigdef_input, tnsr_info in sig_def.inputs.items()}
-            raw_preds = sess.run(fetch_mapping, feed_dict=feed_dict)
-            pred_dict = {fetch_name: list(values) for fetch_name, values in raw_preds.items()}
-            return pandas.DataFrame(data=pred_dict)
+def save_model(tf_saved_model_dir, tf_meta_graph_tags, tf_signature_def_key, path,
+               mlflow_model=Model(), conda_env=None):
+    """
+    Save a *serialized* collection of Tensorflow graphs and variables as an MLflow model
+    to a local path. This method operates on Tensorflow variables and graphs that have been
+    serialized in Tensorflow's `SavedModel` format. For more information about the `SavedModel`,
+    see the following Tensorflow documentation: https://www.tensorflow.org/guide/saved_model#
+    save_and_restore_models.
+
+    :param tf_saved_model_dir: Path to the directory containing serialized Tensorflow variables and
+                               graphs in `SavedModel` format.
+    :param tf_meta_graph_tags: a list of tags identifying the model's metagraph within the
+                               serialized `savedmodel` object. for more information, see the `tags`
+                               parameter of the `tf.saved_model.builder.savedmodelbuilder` method:
+                               https://www.tensorflow.org/api_docs/python/tf/saved_model/builder/
+                               savedmodelbuilder#add_meta_graph
+    :param tf_signature_def_key: a string identifying the input/output signature associated with the
+                                 model. this is a key within the serialized `savedmodel`'s signature
+                                 definition mapping. for more information, see the
+                                 `signature_def_map` parameter of the
+                                 `tf.saved_model.builder.savedmodelbuilder` method.
+    :param path: Local path where the MLflow model is to be saved.
+    :param mlflow_model: MLflow model configuration to which this flavor will be added.
+    :param conda_env: Path to a Conda environment file. If provided, defines an environment for the
+                      model. At minimum, it should specify python, tensorflow, and mlflow with
+                      appropriate versions.
+    """
+    if os.path.exists(path):
+        raise MlflowException("Path '{}' already exists".format(path), DIRECTORY_NOT_EMPTY)
+    os.makedirs(path)
```

Comment:
Could we validate here that the specified `meta_graph_tags`, `signature_def_key` are actually present in the saved model? It'd be nice to do that so users don't persist a model that can't be loaded back properly

Reply:
Yep, good idea! Added a `_validate_saved_model` method that attempts to load the specified Tensorflow model in a fresh graph, emitting useful error output if the loading process fails.